lib/vtls/gtls.c

```diff
@@ -1708,46 +1793,72 @@ static CURLcode
 gtls_connect_common(struct Curl_cfilter *cf,
                     struct Curl_easy *data,
                     bool nonblocking,
-                    bool *done)
-{
+                    bool *done) {
   struct ssl_connect_data *connssl = cf->ctx;
-  CURLcode rc;
+  struct gtls_ssl_backend_data *backend =
+      (struct gtls_ssl_backend_data *)connssl->backend;
   CURLcode result = CURLE_OK;
 
+  DEBUGASSERT(backend);
+
   /* Initiate the connection, if not already done */
-  if(ssl_connect_1 == connssl->connecting_state) {
-    rc = gtls_connect_step1(cf, data);
-    if(rc) {
-      result = rc;
+  if(connssl->connecting_state == ssl_connect_1) {
+    result = gtls_connect_step1(cf, data);
+    if(result)
       goto out;
-    }
+    connssl->connecting_state = ssl_connect_2;
   }
 
-  rc = handshake(cf, data, TRUE, nonblocking);
-  if(rc) {
-    /* handshake() sets its own error message with failf() */
-    result = rc;
-    goto out;
+  if(connssl->connecting_state == ssl_connect_2) {
+    if(connssl->earlydata_state == ssl_earlydata_use) {
+      infof(data, "TLS EarlyData enabled, delaying handshake");
+      connssl->state = ssl_connection_deferred;
+      goto out;
+    }
+    else if(connssl->earlydata_state == ssl_earlydata_sending) {
+      result = gtls_send_earlydata(cf, data);
```

Comment:
Now that the logic is slightly changed to allow sending from `connect_common` there's the problem of the previously agreed upon `alpn` to not be honoured when we're sending early data in protocols like `HTTP/2`, since that's set in `gtls_verifyserver`. Doing this will end up sending a `HTTP/1.1` request after negotiating `h2`.

```
* ALPN: server accepted h2
* Server accpeted 65 bytes of early data
* Curl_xfer_send(len=65, eos=1) -> 0, 65
> GET / HTTP/1.1
```

OpenSSL has an api to query the alpn from a session object, that gnutls equivalent doesn't seem to work.

Reply:
Ah, good point! Hmm...we would need to remember the ALPN in our cache then, most likely. Tricky.